lib/kino/ecto.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Kino.Ecto do
+  @moduledoc """
+  A widget for interactively viewing `Ecto` query results.
+
+  The data must be an enumerable of records, where each
+  record is either map, struct, keyword list or tuple.
+
+  ## Examples
+
+  The widget primarly allows for viewing a database table
+  given a schema:
+
+      Kino.Ecto.new(Weather, Repo)
+
+  However, the first argument can be any queryable, so
+  you can pipe arbitrary queries directly to the widget:
+
+      from(w in Weather, where: w.city == "New York")
+      |> Kino.Ecto.new(Repo)
+  """
+
+  use GenServer, restart: :temporary
+
+  alias Kino.Utils.Table
+
+  defstruct [:pid]
+
+  import Ecto.Query, only: [from: 2]
+
+  @type t :: %__MODULE__{pid: pid()}
+
+  @typedoc false
+  @type state :: %{
+          parent_monitor_ref: reference(),
+          repo: Ecto.Repo.t(),
+          queryable: Ecto.Queryable.t()
+        }
+
+  @doc """
+  Starts a widget process with the given queryable as
+  the data source.
+  """
+  @spec new(Ecto.Queryable.t(), Ecto.Repo.t()) :: t()
+  def new(queryable, repo) when is_atom(repo) do
+    unless queryable?(queryable) do
+      raise ArgumentError,
+            "expected a term implementing the Ecto.Queryable protocol, got: #{inspect(queryable)}"
+    end
+
+    parent = self()
+    opts = [repo: repo, queryable: queryable, parent: parent]
+
+    {:ok, pid} = DynamicSupervisor.start_child(Kino.WidgetSupervisor, {__MODULE__, opts})
+
+    %__MODULE__{pid: pid}
+  end
+
+  defp queryable?(term) do
+    Ecto.Queryable.impl_for(term) != nil
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl true
+  def init(opts) do
+    repo = Keyword.fetch!(opts, :repo)
+    queryable = Keyword.fetch!(opts, :queryable)
+    parent = Keyword.fetch!(opts, :parent)
+
+    parent_monitor_ref = Process.monitor(parent)
+
+    {:ok, %{parent_monitor_ref: parent_monitor_ref, repo: repo, queryable: queryable}}
+  end
+
+  @impl true
+  def handle_info({:connect, pid}, state) do
+    {name, columns} =
+      if Table.ecto_schema?(state.queryable) do
+        name = state.queryable.__schema__(:source) |> to_string()
+        columns = state.queryable.__schema__(:fields) |> Table.keys_to_columns()
+        {name, columns}
+      else
+        {"Query", []}
+      end
+
+    features =
+      Kino.Utils.truthy_keys(
+        refetch: true,
+        pagination: true,
+        # If the user specifies custom select, the record keys
+        # are not valid "order by" fields, so we disable sorting
+        sorting: default_select_query?(state.queryable)
```

Comment:
Should we sort if there is already sorting?

Reply:
I should be fine. We show the initially sorted results, but the user can also click on columns to override the sorting.

Reply:
The issue is that we are not overriding, we are adding to the sorting. We would need to exclude the existing sorting first. We should have tests for this :)

Reply:
Oh right! Exclude sounds good. Is there some test Repo that we could use or how would you go about the tests?

Reply:
There is an ETSO adapter for ETS, no dependencies but I am not sure if it supports order by. Otherwise either PG or mocks.

Reply:
There's an old [pending PR](https://github.com/evadne/etso/pull/12) that adds ORDER BY support ot ETSO. Either way, to make precise assertions we need mocking in some form. I've just introduced a `MockRepo` that we can assert and resolve calls against, let me know if it makes sense :)